README.md
# syscoin-core
A modern Syscoin Core REST and RPC client to execute administrative tasks, wallet operations and queries about network and the blockchain.

## Status
[![npm version][npm-image]][npm-url] [![build status][travis-image]][travis-url]

## Installation
Install the package via `npm`:

```sh
npm install syscoin-core --save
```

## Usage
### Client(...args)
#### Arguments
1. `[agentOptions]` _(Object)_: Optional `agent` [options](https://github.com/request/request#using-optionsagentoptions) to configure SSL/TLS.
2. `[headers=false]` _(boolean)_: Whether to return the response headers.
3. `[host=localhost]` _(string)_: The host to connect to.
4. `[logger=debugnyan('syscoin-core')]` _(Function)_: Custom logger (by default, `debugnyan`).
5. `[network=mainnet]` _(string)_: The network
6. `[password]` _(string)_: The RPC server user password.
7. `[port=[network]]` _(string)_: The RPC server port.
8. `[ssl]` _(boolean|Object)_: Whether to use SSL/TLS with strict checking (_boolean_) or an expanded config (_Object_).
9. `[ssl.enabled]` _(boolean)_: Whether to use SSL/TLS.
10. `[ssl.strict]` _(boolean)_: Whether to do strict SSL/TLS checking (certificate must match host).
11. `[timeout=30000]` _(number)_: How long until the request times out (ms).
12. `[username]` _(number)_: The RPC server user name.
13. `[version]` _(string)_: Which version to check methods for ([read more](#versionchecking)).

### Examples
#### Using network mode
The `network` will automatically determine the port to connect to, just like the `syscoind` and `syscoin-cli` commands.

```js
const Client = require('syscoin-core');
const client = new Client({ network: 'regtest' });
```

##### Setting a custom port

```js
const client = new Client({ port: 28332 });
```

#### Connecting to an SSL/TLS server with strict checking enabled
By default, when `ssl` is enabled, strict checking is implicitly enabled.

```js
const fs = require('fs');
const client = new Client({
  agentOptions: {
    ca: fs.readFileSync('/etc/ssl/syscoind/cert.pem')
  },
  ssl: true
});
```

#### Connecting to an SSL/TLS server without strict checking enabled

```js
const client = new Client({
  ssl: {
    enabled: true,
    strict: false
  }
});
```

#### Using promises to process the response

```js
client.getInfo().then((help) => console.log(help));
```

#### Using callbacks to process the response

```js
client.getInfo((error, help) => console.log(help));
```

#### Returning headers in the response
For compatibility with other Syscoin Core clients.

```js
const client = new Client({ headers: true });

// Promise style with headers enabled:
client.getInfo().then(([body, headers]) => console.log(body, headers));

// Await style based on promises with headers enabled:
const [body, headers] = await client.getInfo();
```

### Floating point number precision in JavaScript

Due to [Javascript's limited floating point precision](http://floating-point-gui.de/), all big numbers (numbers with more than 15 significant digits) are returned as strings to prevent precision loss.

### Version Checking
By default, all methods are exposed on the client independently of the version it is connecting to. This is the most flexible option as defining methods for unavailable RPC calls does not cause any harm and the library is capable of handling a `Method not found` response error correctly.

```js
const client = new Client();

client.command('foobar');
// => RpcError: -32601 Method not found
```

However, if you prefer to be on the safe side, you can enable strict version checking. This will validate all method calls before executing the actual RPC request:

```js
const client = new Client({ version: '0.12.0' });

client.getHashesPerSec();
// => Method "gethashespersec" is not supported by version "0.12.0"
```

If you want to enable strict version checking for the bleeding edge version, you may set a very high version number to exclude recently deprecated calls:

```js
const client = new Client({ version: `${Number.MAX_SAFE_INTEGER}.0.0` });

client.getWork();
// => Throws 'Method "getwork" is not supported by version "9007199254740991.0.0"'.
```

To avoid potential issues with prototype references, all methods are still enumerable on the library client prototype.

### RPC
Start the `syscoind` with the RPC server enabled and optionally configure a username and password:

```sh
docker run --rm -it seegno/syscoind:0.12-alpine -printtoconsole -rpcuser=foo -rpcpassword=bar -server
```

These configuration values may also be set on the `syscoin.conf` file of your platform installation.

By default, port `8332` is used to listen for requests in `mainnet` mode, or `18332` in `testnet` or `regtest` modes. Use the `network` property to initialize the client on the desired mode and automatically set the respective default port. You can optionally set a custom port of your choice too.

The RPC services binds to the localhost loopback network interface, so use `rpcbind` to change where to bind to and `rpcallowip` to whitelist source IP access.

#### Methods
All RPC [methods](src/methods.js) are exposed on the client interface as a camelcase'd version of those available on `syscoind`.

For a more complete reference about which methods are available, check the [RPC documentation](https://syscoin.org/en/developer-reference#remote-procedure-calls-rpcs) on the [Syscoin Core Developer Reference website](https://syscoin.org/en/developer-reference).

##### Examples

```js
client.createRawTransaction([{ txid: '1eb590cd06127f78bf38ab4140c4cdce56ad9eb8886999eb898ddf4d3b28a91d', vout: 0 }], { 'mgnucj8nYqdrPFh2JfZSB1NmUThUGnmsqe': 0.13 });
client.sendMany('test1', { mjSk1Ny9spzU2fouzYgLqGUD8U41iR35QN: 0.1, mgnucj8nYqdrPFh2JfZSB1NmUThUGnmsqe: 0.2 }, 6, 'Example Transaction');
client.sendToAddress('mmXgiR6KAhZCyQ8ndr2BCfEq1wNG2UnyG6', 0.1,  'sendtoaddress example', 'Nemo From Example.com');
```

#### Batch requests
Batched requests are support by passing an array to the `command` method with a `method` and optionally, `parameters`. The return value will be an array with all the responses.

```js
const batch = [
  { method: 'getnewaddress', parameters: [] },
  { method: 'getnewaddress', parameters: [] }
]

new Client().command(batch).then((responses) => console.log(responses)));

// Or, using ES2015 destructuring.
new Client().command(batch).then(([firstAddress, secondAddress]) => console.log(firstAddress, secondAddress)));
```

Note that batched requests will only throw an error if the batch request itself cannot be processed. However, each individual response may contain an error akin to an individual request.

```js
const batch = [
  { method: 'foobar', params: [] },
  { method: 'getnewaddress', params: [] }
]

new Client().command(batch).then(([address, error]) => console.log(address, error)));
// => `mkteeBFmGkraJaWN5WzqHCjmbQWVrPo5X3, { [RpcError: Method not found] message: 'Method not found', name: 'RpcError', code: -32601 }`.
```

### REST
Support for the REST interface is still **experimental** and the API is still subject to change. These endpoints are also **unauthenticated** so [there are certain risks which you should be aware](https://github.com/syscoin/bitcoin/blob/master/doc/REST-interface.md#risks), specifically of leaking sensitive data of the node if not correctly protected.

Error handling is still fragile so avoid passing user input.

Start the `syscoind` with the REST server enabled:

```sh
docker run --rm -it seegno/syscoind:0.12-alpine -printtoconsole -server -rest
```

These configuration values may also be set on the `syscoin.conf` file of your platform installation. Use `txindex=1` if you'd like to enable full transaction query support (note: this will take a considerable amount of time on the first run).

### Methods
#### getBlockByHash(hash, [options], [callback])
Given a block hash, returns a block, in binary, hex-encoded binary or JSON formats.

##### Arguments
1. `hash` _(string)_: The block hash.
2. `[options]` _(Object)_: The options object.
3. `[options.extension=json]` _(string)_: Return in binary (`bin`), hex-encoded binary (`hex`) or JSON (`json`) format.
4. `[callback]` _(Function)_: An optional callback, otherwise a Promise is returned.

##### Example

```js
client.getBlockByHash('0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206', { extension: 'json' });
```

#### getBlockHeadersByHash(hash, count, [options][, callback])
Given a block hash, returns amount of block headers in upward direction.

##### Arguments
1. `hash` _(string)_: The block hash.
2. `count` _(number)_: The number of blocks to count in upward direction.
3. `[options]` _(Object)_: The options object.
4. `[options.extension=json]` _(string)_: Return in binary (`bin`), hex-encoded binary (`hex`) or JSON (`json`) format.
5. `[callback]` _(Function)_: An optional callback, otherwise a Promise is returned.

##### Example

```js
client.getBlockHeadersByHash('0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206', 1, { extension: 'json' });
```

#### getBlockchainInformation([callback])
Returns various state info regarding block chain processing.

#### Arguments
1. `[callback]` _(Function)_: An optional callback, otherwise a Promise is returned.

##### Example

```js
client.getBlockchainInformation([callback]);
```

#### getMemoryPoolContent()
Returns transactions in the transaction memory pool.

#### Arguments
1. `[callback]` _(Function)_: An optional callback, otherwise a Promise is returned.

##### Example

```js
client.getMemoryPoolContent();
```

#### getMemoryPoolInformation([callback])
Returns various information about the transaction memory pool. Only supports JSON as output format.
- size: the number of transactions in the transaction memory pool.
- bytes: size of the transaction memory pool in bytes.
- usage: total transaction memory pool memory usage.

#### Arguments
1. `[callback]` _(Function)_: An optional callback, otherwise a Promise is returned.

##### Example

```js
client.getMemoryPoolInformation();
```

#### getTransactionByHash(hash, [options], [callback])
Given a transaction hash, returns a transaction in binary, hex-encoded binary, or JSON formats.

#### Arguments
1. `hash` _(string)_: The transaction hash.
2. `[options]` _(Object)_: The options object.
3. `[options.summary=false]` _(boolean)_: Whether to return just the transaction hash, thus saving memory.
4. `[options.extension=json]` _(string)_: Return in binary (`bin`), hex-encoded binary (`hex`) or JSON (`json`) format.
5. `[callback]` _(Function)_: An optional callback, otherwise a Promise is returned.

##### Example

```js
client.getTransactionByHash('b4dd08f32be15d96b7166fd77afd18aece7480f72af6c9c7f9c5cbeb01e686fe', { extension: 'json', summary: false });
```

### getUnspentTransactionOutputs(outpoints, [options], [callback])
Query unspent transaction outputs (UTXO) for a given set of outpoints. See [BIP64](https://github.com/bitcoin/bips/blob/master/bip-0064.mediawiki) for input and output serialisation.

#### Arguments
1. `outpoints` _(array\<Object\>|Object)_: The outpoint to query in the format `{ id: '<txid>', index: '<index>' }`.
2. `[options]` _(Object)_: The options object.
3. `[options.extension=json]` _(string)_: Return in binary (`bin`), hex-encoded binary (`hex`) or JSON (`json`) format.
4. `[callback]` _(Function)_: An optional callback, otherwise a Promise is returned.

##### Example

```js
client.getUnspentTransactionOutputs([{
  id: '0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206',
  index: 0
}, {
  id: '0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206',
  index: 1
}], { extension: 'json' }, [callback])
```

### SSL
This client supports SSL out of the box. Simply pass the SSL public certificate to the client and optionally disable strict SSL checking which will bypass SSL validation (the connection is still encrypted but the server it is connecting to may not be trusted). This is, of course, discouraged unless for testing purposes when using something like self-signed certificates.

#### Generating a self-signed certificates for testing purposes
Please note that the following procedure should only be used for testing purposes.

Generate an self-signed certificate together with an unprotected private key:

```sh
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 3650 -nodes
```

#### Connecting via SSL
On Syscoin Core <0.12, you can start the `syscoind` RPC server directly with SSL:

```sh
docker run --rm -it -v $(PWD)/ssl:/etc/ssl seegno/syscoind:0.11-alpine -printtoconsole -rpcuser=foo -rpcpassword=bar -rpcssl -rpcsslcertificatechainfile=/etc/ssl/syscoind/cert.pem -rpcsslprivatekeyfile=/etc/ssl/syscoind/key.pem -server
```

On Syscoin Core >0.12, use must use `stunnel` (`brew install stunnel` or `sudo apt-get install stunnel4`) or an HTTPS reverse proxy to configure SSL since the built-in support for SSL has been removed. The trade off with `stunnel` is performance and simplicity versus features, as it lacks more powerful capacities such as Basic Authentication and caching which are standard in reverse proxies.

You can use `stunnel` by configuring `stunnel.conf` with the following service requirements:

```
[syscoin]
accept = 28332
connect = 18332
cert = /etc/ssl/syscoind/cert.pem
key = /etc/ssl/syscoind/key.pem
```

The `key` option may be omitted if you concatenating your private and public certificates into a single `stunnel.pem` file.

On some versions of `stunnel` it is also possible to start a service using command line arguments. The equivalent would be:

```sh
stunnel -d 28332 -r 127.0.0.1:18332 -p stunnel.pem -P ''
```

Then pass the public certificate to the client:

```js
const Client = require('syscoin-core');
const fs = require('fs');
const client = new Client({
  agentOptions: {
    ca: fs.readFileSync('/etc/ssl/syscoind/cert.pem')
  },
  port: 28332,
  ssl: true
});
```

## Logging

By default, all requests made with `syscoin-core` are logged using [uphold/debugnyan](https://github.com/uphold/debugnyan) with `syscoin-core` as the logging namespace.

Please note that all sensitive data is obfuscated before calling the logger.

#### Example

Example output defining the environment variable `DEBUG=syscoin-core`:

```javascript
const client = new Client();

client.getTransactionByHash('b4dd08f32be15d96b7166fd77afd18aece7480f72af6c9c7f9c5cbeb01e686fe');

// {
//   "name": "syscoin-core",
//   "hostname": "localhost",
//   "pid": 57908,
//   "level": 20,
//   "request": {
//     "headers": {
//       "host": "localhost:8332",
//       "accept": "application/json"
//     },
//     "id": "82cea4e5-2c85-4284-b9ec-e5876c84e67c",
//     "method": "GET",
//     "type": "request",
//     "uri": "http://localhost:8332/rest/tx/b4dd08f32be15d96b7166fd77afd18aece7480f72af6c9c7f9c5cbeb01e686fe.json"
//   },
//   "msg": "Making request 82cea4e5-2c85-4284-b9ec-e5876c84e67c to GET http://localhost:8332/rest/tx/b4dd08f32be15d96b7166fd77afd18aece7480f72af6c9c7f9c5cbeb01e686fe.json",
//   "time": "2017-02-07T14:40:35.020Z",
//   "v": 0
// }
```

### Custom logger

A custom logger can be passed via the `logger` option and it should implement [bunyan's log levels](https://github.com/trentm/node-bunyan#levels).

## Tests
Currently the test suite is tailored for Docker (including `docker-compose`) due to the multitude of different `syscoind` configurations that are required in order to get the test suite passing.

To test using a local installation of `node.js` but with dependencies (e.g. `syscoind`) running inside Docker:

```sh
npm run dependencies
npm test
```

To test using Docker exclusively (similarly to what is done in Travis CI):

```sh
npm run testdocker
```

## Release

```sh
npm version [<newversion> | major | minor | patch] -m "Release %s"
```

## License
MIT

[npm-image]: https://img.shields.io/npm/v/bitcoin-core.svg?style=flat-square
[npm-url]: https://npmjs.org/package/syscoin-core
[travis-image]: https://img.shields.io/travis/ruimarinho/bitcoin-core.svg?style=flat-square
[travis-url]: https://travis-ci.org/ruimarinho/syscoin-core


CHANGELOG.md
# Changelog

## [3.0.0-beta](https://github.com/syscoin/syscoin-core/tree/3.0.0-beta) (2018-06-02)
[Full Changelog](https://github.com/syscoin/syscoin-core/compare/v1.2.0...3.0.0-beta)

**Merged pull requests:**

- Name change from aliasnewfund -\> systxfund [\#7](https://github.com/syscoin/syscoin-core/pull/7) ([justicegray](https://github.com/justicegray))
- Feature/asset allocation collect interest [\#6](https://github.com/syscoin/syscoin-core/pull/6) ([justicegray](https://github.com/justicegray))
- BD-273 BD-274 BD-275  messages 3.0 [\#5](https://github.com/syscoin/syscoin-core/pull/5) ([dwasyluk](https://github.com/dwasyluk))
- Bd 173 signrawtransaction [\#4](https://github.com/syscoin/syscoin-core/pull/4) ([quangogster](https://github.com/quangogster))
- Fixes to syscoinlistreceivedbyaddress [\#3](https://github.com/syscoin/syscoin-core/pull/3) ([dwasyluk](https://github.com/dwasyluk))
- added syscoinlistreceivebyaddress [\#2](https://github.com/syscoin/syscoin-core/pull/2) ([quangogster](https://github.com/quangogster))
- Update methods [\#1](https://github.com/syscoin/syscoin-core/pull/1) ([quangogster](https://github.com/quangogster))

## [v1.2.0](https://github.com/syscoin/syscoin-core/tree/v1.2.0) (2017-02-12)
[Full Changelog](https://github.com/syscoin/syscoin-core/compare/v1.1.0...v1.2.0)

## [v1.1.0](https://github.com/syscoin/syscoin-core/tree/v1.1.0) (2016-09-19)
[Full Changelog](https://github.com/syscoin/syscoin-core/compare/v1.0.0...v1.1.0)



src/methods.js
/* eslint-disable no-inline-comments,sort-keys */
import { set } from 'lodash';

/**
 * Export available rpc methods.
 */

/**
 * @Dev
 * follow the versioning rule below
 * to update changes:
 * *new as of version : '>= new_version'
 * *removed as of version: '< new_version'
 * *existed for version range: '>= start_version < end_version'
 */

export default {
  addMultiSigAddress: { version: '>=2.2.0' },
  aliasAddScript: { version: '>=2.2.0' },
  aliasBalance: { version: '>=2.2.0' },
  aliasClearWhitelist: { version: '>=2.2.0' },
  aliasCount: { version: '>=2.2.0 <3.0.0' },
  aliasFilter: { version: '>=2.2.0' },
  aliasHistory: { version: '>=2.2.0' },
  aliasInfo: { version: '>=2.2.0' },
  aliasList: { version: '>=2.2.0 <3.0.0' },
  aliasNew: { version: '>=2.2.0' },
  aliasPay: { version: '>=3.0.0' },
  aliasTxHistory: { version: '>=2.2.0' },
  aliasUpdate: { version: '>=2.2.0' },
  aliasUpdateWhitelist: { version: '>=2.2.0' },
  aliasWhitelist: { version: '>=2.2.0' },
  assetAllocationCollectInterest: { version: '>=3.0.0' },
  assetAllocationInfo: { version: '>=3.0.0' },
  assetAllocationSend: { version: '>=3.0.0' },
  assetAllocationSenderStatus: { version: '>=3.0.0' },
  assetInfo: { version: '>=3.0.0' },
  assetNew: { version: '>=3.0.0' },
  assetSend: { version: '>=3.0.0' },
  assetTransfer: { version: '>=3.0.0' },
  assetUpdate: { version: '>=3.0.0' },
  certFilter: { version: '>=2.2.0' },
  certHistory: { version: '>=2.2.0' },
  certInfo: { version: '>=2.2.0' },
  certNew: { version: '>=2.2.0' },
  certTransfer: { version: '>=2.2.0' },
  certUpdate: { version: '>=2.2.0' },
  debug: { version: '>=3.0.0' },
  dumpHdInfo: { version: '>=3.0.0' },
  dumpPrivKey: {
    obfuscate: {
      response: () => '******'
    },
    version: '>=2.2.0'
  },
  dumpWallet: { version: '>=2.2.0' },
  encryptWallet: {
    obfuscate: {
      request: params => set([...params], '[0]', '******')
    },
    version: '>=2.2.0'
  },
  escrowAcknowledge: { version: '>=2.2.0' },
  escrowAddShipping: { version: '>=2.2.0' },
  escrowBid: { version: '>=2.2.0' },
  escrowBidHistory: { version: '>=2.2.0' },
  escrowCompleteRefund: { version: '>=2.2.0' },
  escrowCompleteRelease: { version: '>=2.2.0' },
  escrowCreateRawTransaction: { version: '>=2.2.0' },
  escrowFeedback: { version: '>=2.2.0' },
  escrowFeedbackHistory: { version: '>=2.2.0' },
  escrowFilter: { version: '>=2.2.0' },
  escrowInfo: { version: '>=2.2.0' },
  escrowNew: { version: '>=2.2.0' },
  escrowRefund: { version: '>=2.2.0' },
  escrowRelease: { version: '>=2.2.0' },
  fundRawTransaction: { version: '>=3.0.0' },
  getBlock: { version: '>=2.2.0' },
  getBlockCount: { version: '>=2.2.0' },
  getBlockchainInfo: { version: '>=2.2.0' },
  getInfo: { version: '>=2.2.0' },
  getMiningInfo: { version: '>=2.2.0' },
  getNetworkInfo: { version: '>=2.2.0' },
  getPeerInfo: { version: '>=2.2.0' },
  pruneSyscoinServices: { version: '>=2.2.0' },
  signRawTransaction: { version: '>=2.2.0' },
  syscoinDecodeRawTransaction: { version: '>=2.2.0' },
  syscoinListReceivedByAddress: { version: '>=3.0.0' },
  syscoinQuery: { version: '>=2.2.0' },
  syscoinSendRawTransaction: { version: '>=2.2.0' },
  verifyMessage: { version: '>=2.2.0' },
  getAddressDeltas: { version: '>=3.0.0' },
  getAccount: { version: '>=2.2.0' },
  getAccountAddress: { version: '>=2.2.0' },
  getAddressBalance: { version: '>=2.2.0' },
  getAddressUtxos: { version: '>=2.2.0' },
  getAddressMempool: { version: '>=3.0.0' },
  getAddressTxids: { version: '>=3.0.0' },
  getBlockHashes: { version: '>=3.0.0' },
  getBlockHeaders: { version: '>=3.0.0' },
  getChainTips: { version: '>=3.0.0' },
  getSpentInfo: { version: '>=3.0.0' },
  getGenerate: { version: '>=3.0.0' },
  getGovernanceInfo: { version: '>=3.0.0' },
  getPoolInfo: { version: '>=3.0.0' },
  getSuperBlockBudget: { version: '>=3.0.0' },
  getBalance: { version: '>=2.2.0' },
  getWalletBalance: { version: '>=2.2.0' },
  getNewAddress: { version: '>=2.2.0' },
  getReceivedByAccount: { version: '>=2.2.0' },
  getReceivedByAddress: { version: '>=2.2.0' },
  getTransaction: { version: '>=2.2.0' },
  getUnconfirmedBalance: { version: '>=2.2.0' },
  getWalletInfo: { version: '>=2.2.0' },
  getBlockTemplate: { version: '>=3.0.0' },
  gObject: { version: '>=3.0.0' },
  importElectrumWallet: { version: '>=3.0.0' },
  importAddress: { version: '>=2.2.0' },
  importPrivKey: {
    obfuscate: {
      request: () => ['******']
    },
    version: '>=2.2.0'
  },
  importPubKey: { version: '>=2.2.0' },
  importWallet: { version: '>=2.2.0' },
  instantSendToAddress: { version: '>=3.0.0' },
  keepass: { version: '>=3.0.0' },
  lockUnspent: { version: '>=3.0.0' },
  listAccounts: { version: '>=2.2.0' },
  listAddressGroupings: { version: '>=2.2.0' },
  listReceivedByAccount: { version: '>=2.2.0' },
  listReceivedByAddress: { version: '>=2.2.0' },
  listSinceBlock: { version: '>=2.2.0' },
  listTransactions: { version: '>=2.2.0' },
  listUnspent: { version: '>=3.0.0' },
  masternode: { version: '>=3.0.0' },
  masternodeBroadcast: { version: '>=3.0.0' },
  masternodeList: { version: '>=3.0.0' },
  mnSync: { version: '>=3.0.0' },
  offerInfo: { version: '>=2.2.0' },
  offerHistory: { version: '>=2.2.0' },
  offerFilter: { version: '>=2.2.0' },
  offerLink: { version: '>=2.2.0' },
  offerNew: { version: '>=2.2.0' },
  offerUpdate: { version: '>=2.2.0' },
  privateSend: { version: '>=3.0.0' },
  sentinelPing: { version: '>=3.0.0' },
  spork: { version: '>=3.0.0' },
  sendRawTransaction: { version: '>=3.0.0' },
  setGenerate: { version: '>=3.0.0' },
  setNetworkActive: { version: '>=3.0.0' },
  sendFrom: { version: '>=2.2.0' },
  sendMany: { version: '>=2.2.0' },
  sendToAddress: { version: '>=2.2.0' },
  signMessage: { version: '>=2.2.0' },
  syscoinListReceiveByAddress: { version: '>=3.0.0' },
  syscoinTxFund: { version: '>=3.0.0' },
  validateAddress: { version: '>=2.2.0' },
  voteRaw: { version: '>=3.0.0' },
  walletLock: { version: '>=2.2.0' },
  walletPassphrase: {
    obfuscate: {
      request: params => set([...params], '[0]', '******')
    },
    version: '>=2.2.0'
  },
  walletPassphraseChange: { version: '>=2.2.0' }
};


src/index.js

/**
 * Module dependencies.
 */

import Parser from './parser';
import Promise from 'bluebird';
import Requester from './requester';
import _ from 'lodash';
import bunyan from 'bunyan';
import methods from './methods';
import requestLogger from './logging/request-logger';
import semver from 'semver';

/**
 * Source arguments to find out if a callback has been passed.
 */

function source(...args) {
  const last = _.last(args);

  let callback;

  if (_.isFunction(last)) {
    callback = last;
    args = _.dropRight(args);
  }

  return [args, callback];
}

/**
 * List of networks and their default port mapping.
 */

const networks = {
  mainnet: 8332,
  regtest: 18332,
  testnet: 18332
};

/**
 * Constructor.
 */

class Client {
  constructor({
    agentOptions,
    headers = false,
    host = 'localhost',
    logger = bunyan({ name: 'syscoin-core' }),
    network = 'mainnet',
    password,
    port,
    ssl = false,
    timeout = 30000,
    username,
    version,
    sysversion
  } = {}) {
    if (!_.has(networks, network)) {
      throw new Error(`Invalid network name "${network}"`, { network });
    }

    this.agentOptions = agentOptions;
    this.auth = (password || username) && { pass: password, user: username };
    this.headers = headers;
    this.host = host;
    this.password = password;
    this.port = port || networks[network];
    this.timeout = timeout;
    this.ssl = {
      enabled: _.get(ssl, 'enabled', ssl),
      strict: _.get(ssl, 'strict', _.get(ssl, 'enabled', ssl))
    };

    // Find unsupported methods according to version and sysversion.
    let unsupported = [];

    if (version || sysversion) {
      unsupported = _.chain(methods)
        .pickBy(method => {
          let pick = false;

          if (version) {
            pick = !semver.satisfies(version, method.version);
          }
          if (sysversion) {
            pick = !semver.satisfies(sysversion, method.sysversion);
          }

          return pick;
        })
        .keys()
        .invokeMap(String.prototype.toLowerCase)
        .value();
    }

    const request = requestLogger(logger);

    this.request = Promise.promisifyAll(request.defaults({
      agentOptions: this.agentOptions,
      baseUrl: `${this.ssl.enabled ? 'https' : 'http'}://${this.host}:${this.port}`,
      json: true,
      strictSSL: this.ssl.strict,
      timeout: this.timeout
    }), { multiArgs: true });
    this.requester = new Requester({ sysversion, unsupported, version });
    this.parser = new Parser({ headers: this.headers });
  }

  /**
   * Execute `rpc` command.
   */

  command(...args) {
    let body;
    let callback;
    let parameters = _.tail(args);
    const input = _.head(args);
    const lastArg = _.last(args);

    if (_.isFunction(lastArg)) {
      callback = lastArg;
      parameters = _.dropRight(parameters);
    }

    return Promise.try(() => {
      if (Array.isArray(input)) {
        body = input.map((method, index) => this.requester.prepare({ method: method.method, parameters: method.parameters, suffix: index }));
      } else {
        body = this.requester.prepare({ method: input, parameters });
      }

      return this.request.postAsync({
        auth: _.pickBy(this.auth, _.identity),
        body: JSON.stringify(body),
        json: false,
        uri: '/'
      }).bind(this)
        .then(this.parser.rpc);
    }).asCallback(callback);
  }

  /**
   * Given a transaction hash, returns a transaction in binary, hex-encoded binary, or JSON formats.
   */

  getTransactionByHash(...args) {
    const [[hash, { extension = 'json' } = {}], callback] = source(...args);

    return Promise.try(() => {
      return this.request.getAsync({
        encoding: extension === 'bin' ? null : undefined,
        url: `/rest/tx/${hash}.${extension}`
      }).bind(this)
        .then(this.parser.rest);
    }).asCallback(callback);
  }

  /**
   * Given a block hash, returns a block, in binary, hex-encoded binary or JSON formats.
   * With `summary` set to `false`, the JSON response will only contain the transaction
   * hash instead of the complete transaction details. The option only affects the JSON response.
   */

  getBlockByHash(...args) {
    const [[hash, { summary = false, extension = 'json' } = {}], callback] = source(...args);

    return Promise.try(() => {
      return this.request.getAsync({
        encoding: extension === 'bin' ? null : undefined,
        url: `/rest/block${summary ? '/notxdetails/' : '/'}${hash}.${extension}`
      }).bind(this)
        .then(this.parser.rest);
    }).asCallback(callback);
  }

  /**
   * Given a block hash, returns amount of blockheaders in upward direction.
   */

  getBlockHeadersByHash(...args) {
    const [[hash, count, { extension = 'json' } = {}], callback] = source(...args);

    return Promise.try(() => {
      if (!_.includes(['bin', 'hex'], extension)) {
        throw new Error(`Extension "${extension}" is not supported`);
      }

      return this.request.getAsync({
        encoding: extension === 'bin' ? null : undefined,
        url: `/rest/headers/${count}/${hash}.${extension}`
      }).bind(this)
        .then(this.parser.rest);
    }).asCallback(callback);
  }

  /**
   * Returns various state info regarding block chain processing.
   * Only supports JSON as output format.
   */

  getBlockchainInformation(...args) {
    const [, callback] = source(...args);

    return this.request.getAsync(`/rest/chaininfo.json`)
      .bind(this)
      .then(this.parser.rest)
      .asCallback(callback);
  }

  /**
   * Query unspent transaction outputs for a given set of outpoints.
   * See BIP64 for input and output serialisation:
   * 	 - https://github.com/bitcoin/bips/blob/master/bip-0064.mediawiki
   */

  getUnspentTransactionOutputs(...args) {
    const [[outpoints, { extension = 'json' } = {}], callback] = source(...args);

    const sets = _.flatten([outpoints]).map(outpoint => {
      return `${outpoint.id}-${outpoint.index}`;
    }).join('/');

    return this.request.getAsync({
      encoding: extension === 'bin' ? null : undefined,
      url: `/rest/getutxos/checkmempool/${sets}.${extension}`
    }).bind(this)
      .then(this.parser.rest)
      .asCallback(callback);
  }

  /**
   * Returns transactions in the transaction memory pool.
   * Only supports JSON as output format.
   */

  getMemoryPoolContent(...args) {
    const [, callback] = source(...args);

    return this.request.getAsync('/rest/mempool/contents.json')
      .bind(this)
      .then(this.parser.rest)
      .asCallback(callback);
  }

  /**
   * Returns various information about the transaction memory pool.
   * Only supports JSON as output format.
   *
   *   - size: the number of transactions in the transaction memory pool.
   *   - bytes: size of the transaction memory pool in bytes.
   *   - usage: total transaction memory pool memory usage.
   */

  getMemoryPoolInformation(...args) {
    const [, callback] = source(...args);

    return this.request.getAsync('/rest/mempool/info.json')
      .bind(this)
      .then(this.parser.rest)
      .asCallback(callback);
  }
}

/**
 * Add all known RPC methods.
 */

_.forOwn(methods, (range, method) => {
  Client.prototype[method] = _.partial(Client.prototype.command, method.toLowerCase());
});

/**
 * Export Client class.
 */

export default Client;
